stage0/run.go

```diff
@@ -268,120 +232,77 @@ func lockDir(dir string) error {
 	return os.Setenv(envLockFd, fmt.Sprintf("%v", fd))
 }
 
-func untarRootfs(r io.Reader, dir string) error {
-	tr := tar.NewReader(r)
-	if err := os.MkdirAll(dir, 0776); err != nil {
-		return fmt.Errorf("error creating stage1 rootfs directory: %v", err)
+// setupAppImage attempts to load the app image by the given hash from the store,
+// verifies that the image matches the hash, and extracts the image into a
+// directory in the given dir.
+// It returns the ImageManifest that the image contains.
+// TODO(jonboulle): tighten up the Hash type here; currently it is partially-populated (i.e. half-length sha512)
+func setupAppImage(cfg Config, img types.Hash, cdir string) (*schema.ImageManifest, error) {
+	log.Println("Loading image", img.String())
+
+	ad := common.AppImagePath(cdir, img)
+	err := os.MkdirAll(ad, 0776)
+	if err != nil {
+		return nil, fmt.Errorf("error creating image directory: %v", err)
 	}
 
-	if err := ptar.ExtractTar(tr, dir, false, nil); err != nil {
-		return fmt.Errorf("error extracting rootfs: %v", err)
+	if err := expandImage(cfg, img, ad); err != nil {
+		return nil, fmt.Errorf("error expanding app image: %v", err)
 	}
-	return nil
-}
 
-// unpackRootfs unpacks a stage1 rootfs (compressed file, pointed to by rfs)
-// into dir, returning any error encountered
-func unpackRootfs(rfs string, dir string) error {
-	fh, err := os.Open(rfs)
+	err = os.MkdirAll(filepath.Join(ad, "rootfs/tmp"), 0777)
```

Comment:
I am drawing a total blank, why do we create this dir? (I realise you didn't add it in this diff)